core/src/main/scala/io/projectglow/vcf/BigVCFDatasource.scala

```diff
@@ -58,6 +62,40 @@ object BigVCFDatasource extends HlsUsageLogging {
         .contains(VCFFileWriter.INFER_HEADER))) {
       throw new SparkException("Cannot infer header for empty VCF.")
     }
+
+    val extraInferredOptions = if (options.getOrElse(
+        VCFFileWriter.VCF_HEADER_KEY,
+        VCFFileWriter.INFER_HEADER) == VCFFileWriter.INFER_HEADER) {
+      // If we have to infer the header (explicitly or implicitly), determine all of the samples in the DataFrame.
+      // The samples that will be written consist of the distinct set of sample IDs identified across the DataFrame,
+      // in addition to missing sample IDs. The number of missing sample IDs will be equal to the maximum number
+      // found in any given row.
+      import data.sparkSession.implicits._
+      val distinctSampleLists =
+        data
+          .select("genotypes.sampleId")
+          .as[Array[String]]
+          .distinct()
+          .collect
+      val maxNumMissingSamples = distinctSampleLists.map { sl =>
+        sl.count(_ == null)
+      }.max
+      val inferredSamples = distinctSampleLists
+        .flatten
+        .distinct
+        .dropWhile(_ == null)
+        .sorted
+      val maybeInferredSamples = if (inferredSamples.length > 0) {
```

Comment:
This should be an explicit argument to the stream writer, not passed through the option map.